### PR TITLE
Remove `ValidPos` type

### PR DESCRIPTION
**Pull request by simsurace**
This is an attempt at addressing https://github.com/JuliaDynamics/Agents.jl/issues/1094.
There are remaining ambiguities that need to be resolved:
```julia
Tuple{Method, Method}[
    (
        add_agent!(model::AgentBasedModel, args::Vararg{Any, N}; kwargs...) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:260,
        add_agent!(pos, A::Union{Function, Type}, model::AgentBasedModel, args::Vararg{Any, N}; kwargs...) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:281
    ),
    (
        add_agent!(agent::AbstractAgent, pos, model::AgentBasedModel) @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:220,
        add_agent!(pos, A::Union{Function, Type}, model::AgentBasedModel, args::Vararg{Any, N}; kwargs...) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:281
    ),
    (
        kwcall(::NamedTuple, ::typeof(add_agent!), model::AgentBasedModel, args::Vararg{Any, N}) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:260,
        kwcall(::NamedTuple, ::typeof(add_agent!), pos, model::AgentBasedModel, args::Vararg{Any, N}) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:270
    ), 
    (
        add_agent!(agent::AbstractAgent, pos, model::AgentBasedModel) @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:220,
        add_agent!(pos, model::AgentBasedModel, args::Vararg{Any, N}; kwargs...) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:270
    ),
    (
        kwcall(::NamedTuple, ::typeof(add_agent!), model::AgentBasedModel, args::Vararg{Any, N}) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:260,
        kwcall(::NamedTuple, ::typeof(add_agent!), pos, A::Union{Function, Type}, model::AgentBasedModel, args::Vararg{Any, N}) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:281), (add_agent!(model::AgentBasedModel, args::Vararg{Any, N}; kwargs...) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:260, add_agent!(pos, model::AgentBasedModel, args::Vararg{Any, N}; kwargs...) where N @ Agents ~/.julia/dev/Agents/src/core/space_interaction_API.jl:270
    )
]

```

### CODE REVIEW

**Comment by Datseris** [comment]
yeah as I mentioned on Slack, `add_agent!` is the main problem. I don't say a way to resolve these ambiguities at the moment.

File: src/spaces/grid_general.jl [hunk 1]
```diff
@@ -69,14 +71,14 @@ function calculate_hyperrectangle(space::AbstractGridSpace{D,P}, r) where {D,P}
...
     else
         odd_s, half_s = space_size .% 2, space_size .÷ 2
         r_dims_P = min.(r, half_s)
         r_dims_notP = min.(r, space_size)
-        from_to = (P[i] ? 
-                    (-r_dims_P[i]:r_dims_P[i]-(r_dims_P[i] == half_s[i] && odd_s[i] == 0)) : 
-                    (-r_dims_notP[i]:r_dims_notP[i]) for i in 1:D)
+        from_to = (P[i] ?
+                   (-r_dims_P[i]:r_dims_P[i]-(r_dims_P[i]==half_s[i]&&odd_s[i]==0)) :
+                   (-r_dims_notP[i]:r_dims_notP[i]) for i in 1:D)
```
Comment (Datseris) [inline src/spaces/grid_general.jl:81]: Can you please remove all formatting changes you did in this PR? I also find them unhelpful: removing all these spaces makes the code less readable.

**Comment by Datseris** [comment]
thanks for the PR @simsurace . I would suggest that we use this PR as a general "improve experience of implementing new space downstream".

File: src/spaces/grid_general.jl (hunk 1 above)
Comment (simsurace) [reply src/spaces/grid_general.jl]: Oops, did not mean to change that.

Comment (simsurace) [reply src/spaces/grid_general.jl]: They should be gone now.

**Comment by Tortar** [comment]
These ambiguities shouldn't be hard to solve, they are mainly innocuous ones if I read them correctly...strangely tests seem stuck though

**Comment by Tortar** [comment]
I think we should extend any function related to space that way, actually we could maybe use f(..., space, model) instead of f(..., space) if there are functions which require the model for API coherency

**Comment by simsurace** [comment]
To me it's not clear how to resolve ambiguities in a way that does not break the previous extension protocol for new spaces.
One could just define one top-level method that includes all conflicting signatures and has logic flow based on their types. This would still be lowered to the same thing more or less, but is not very nice in terms of readability.
If someone wants to give a shot, feel free to continue the PR.

**Comment by Datseris** [comment]
to me its not clear either how to resolve the `add_agent!` ambiguities. I don't think it is possible actually. I think we need to redesign `add_agent!` for it to be possible. Or to redesign the  whole concept of `ValidPos`.

@simsurace , out of curiosity, what prevents you from adding your space directly to this repo?

**Comment by Tortar** [comment]
I will give it a shot, let's see

**Comment by Tortar** [comment]
I fixed all the remaining ambiguities, they were all invalid signatures, so they were fortunately innocuous as I initially thought

**Comment by simsurace** [comment]
> @simsurace , out of curiosity, what prevents you from adding your space directly to this repo?

It is supposed to be possible to define new user-defined spaces. While some spaces a user may want to define may be of general interest and deserve upstreaming, I'm sure there are alot that do not warrent being added to and supported by Agents.jl.

**Review by Tortar — APPROVED** [review]

**Comment by Tortar** [comment]
Thank you a lot @simsurace for starting this, it was clearly something really needed

**Pull request merged**